{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene sorting by Chromosome location (TreeMap)\n",
    "The main idea of this paper is to rearrange gene expression profiles data into tree-map images, according to the chromosome each gene belongs to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Loading and Cleaning\n",
    "\n",
    "Data needed to generate these tree-map images are:\n",
    "\n",
    "* Gene expression profiles of cancer patients: These are obtained in '../TCGA_Data_obtaining.ipynb' notebook and preprocessed by MAD measure to maintain the 20k most significative genes.\n",
    "\n",
    "* Chromosomic information of genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene expression profiles of cancer patients\n",
    "brca_ex = pd.read_hdf(\"/mnt/ext/almacen/comun/Francis/almacen/DeepLearning-Bioinformatics/data/PanCancer/mad_filter_pancan_all_TCGA_20.h5\", key = \"brca\")\n",
    "non_brca_ex = pd.read_hdf(\"/mnt/ext/almacen/comun/Francis/almacen/DeepLearning-Bioinformatics/data/PanCancer/mad_filter_pancan_all_TCGA_20.h5\", key = \"non_brca\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-e8dbc6c3e46f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Chromosomic information of genes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m gene_chrom = pd.read_table('/mnt/ext/almacen/comun/Francis/almacen//GenSig/data/ensembl_gene_chromosome.tsv',\n\u001b[0m\u001b[1;32m      3\u001b[0m                            \u001b[0mindex_col\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m                            \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m                            \u001b[0mnames\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene_id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene_id_version'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'chromosome'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene_start'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene_end'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'transcript_start'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'transcript_end'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "# Chromosomic information of genes\n",
    "gene_chrom = pd.read_table('/mnt/ext/almacen/comun/Francis/almacen//GenSig/data/ensembl_gene_chromosome.tsv',\n",
    "                           index_col=0, \n",
    "                           header = 0,\n",
    "                           names=['gene_id', 'gene_id_version', 'chromosome', 'gene_start', 'gene_end', 'transcript_start', 'transcript_end'],\n",
    "                           dtype = {'Gene stable ID': str,\n",
    "                                    'Gene stable ID version': str,\n",
    "                                    'Chromosome/scaffold name': str,\n",
    "                                    'Gene start (bp)': np.int32,\n",
    "                                    'Gene end (bp)': np.int32,\n",
    "                                    'Transcript start (bp)': np.int32,\n",
    "                                    'Transcript end (bp)': np.int32})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_chrom = gene_chrom[~gene_chrom.index.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(64914, 6)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_chrom.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "common = set(gene_chrom.gene_id_version).intersection(brca_ex.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_chrom = gene_chrom.loc[gene_chrom['gene_id_version'].isin(list(common))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     1644\n",
       "19    1136\n",
       "2     1132\n",
       "17    1069\n",
       "12     966\n",
       "3      896\n",
       "16     837\n",
       "6      833\n",
       "7      832\n",
       "11     825\n",
       "5      818\n",
       "15     657\n",
       "4      655\n",
       "8      641\n",
       "10     629\n",
       "9      629\n",
       "14     619\n",
       "X      518\n",
       "22     457\n",
       "20     423\n",
       "13     329\n",
       "18     280\n",
       "21     253\n",
       "Y       19\n",
       "MT      14\n",
       "Name: chromosome, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_chrom['chromosome'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_id_version', 'chromosome', 'gene_start', 'gene_end',\n",
       "       'transcript_start', 'transcript_end'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_chrom.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_chrom.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_version</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000277209</th>\n",
       "      <td>ENSG00000277209.1</td>\n",
       "      <td>14</td>\n",
       "      <td>20343075</td>\n",
       "      <td>20343407</td>\n",
       "      <td>20343075</td>\n",
       "      <td>20343407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000266274</th>\n",
       "      <td>ENSG00000266274.2</td>\n",
       "      <td>17</td>\n",
       "      <td>30959564</td>\n",
       "      <td>30959850</td>\n",
       "      <td>30959564</td>\n",
       "      <td>30959850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000263432</th>\n",
       "      <td>ENSG00000263432.2</td>\n",
       "      <td>5</td>\n",
       "      <td>123022487</td>\n",
       "      <td>123022783</td>\n",
       "      <td>123022487</td>\n",
       "      <td>123022783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239494</th>\n",
       "      <td>ENSG00000239494.2</td>\n",
       "      <td>1</td>\n",
       "      <td>169859756</td>\n",
       "      <td>169860052</td>\n",
       "      <td>169859756</td>\n",
       "      <td>169860052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212232</th>\n",
       "      <td>ENSG00000212232.1</td>\n",
       "      <td>20</td>\n",
       "      <td>17962710</td>\n",
       "      <td>17962946</td>\n",
       "      <td>17962710</td>\n",
       "      <td>17962946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000211459</th>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "      <td>648</td>\n",
       "      <td>1601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000210082</th>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "      <td>1671</td>\n",
       "      <td>3229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198888</th>\n",
       "      <td>ENSG00000198888.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198763</th>\n",
       "      <td>ENSG00000198763.3</td>\n",
       "      <td>MT</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198804</th>\n",
       "      <td>ENSG00000198804.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228253</th>\n",
       "      <td>ENSG00000228253.1</td>\n",
       "      <td>MT</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "      <td>8366</td>\n",
       "      <td>8572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198899</th>\n",
       "      <td>ENSG00000198899.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>8527</td>\n",
       "      <td>9207</td>\n",
       "      <td>8527</td>\n",
       "      <td>9207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198938</th>\n",
       "      <td>ENSG00000198938.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "      <td>9207</td>\n",
       "      <td>9990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198840</th>\n",
       "      <td>ENSG00000198840.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "      <td>10059</td>\n",
       "      <td>10404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243339</th>\n",
       "      <td>ENSG00000243339.3</td>\n",
       "      <td>3</td>\n",
       "      <td>196399911</td>\n",
       "      <td>196400207</td>\n",
       "      <td>196399911</td>\n",
       "      <td>196400207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142530</th>\n",
       "      <td>ENSG00000142530.10</td>\n",
       "      <td>19</td>\n",
       "      <td>50466785</td>\n",
       "      <td>50476753</td>\n",
       "      <td>50466785</td>\n",
       "      <td>50475485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000086967</th>\n",
       "      <td>ENSG00000086967.9</td>\n",
       "      <td>19</td>\n",
       "      <td>50432903</td>\n",
       "      <td>50466321</td>\n",
       "      <td>50432903</td>\n",
       "      <td>50466321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142552</th>\n",
       "      <td>ENSG00000142552.7</td>\n",
       "      <td>19</td>\n",
       "      <td>49527618</td>\n",
       "      <td>49546962</td>\n",
       "      <td>49527618</td>\n",
       "      <td>49543638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274029</th>\n",
       "      <td>ENSG00000274029.1</td>\n",
       "      <td>12</td>\n",
       "      <td>121190868</td>\n",
       "      <td>121191518</td>\n",
       "      <td>121190868</td>\n",
       "      <td>121191518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000112379</th>\n",
       "      <td>ENSG00000112379.8</td>\n",
       "      <td>6</td>\n",
       "      <td>138161921</td>\n",
       "      <td>138344663</td>\n",
       "      <td>138161921</td>\n",
       "      <td>138344663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228150</th>\n",
       "      <td>ENSG00000228150.1</td>\n",
       "      <td>1</td>\n",
       "      <td>9942923</td>\n",
       "      <td>9949974</td>\n",
       "      <td>9942923</td>\n",
       "      <td>9949974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000233705</th>\n",
       "      <td>ENSG00000233705.6</td>\n",
       "      <td>7</td>\n",
       "      <td>107653968</td>\n",
       "      <td>107662151</td>\n",
       "      <td>107653968</td>\n",
       "      <td>107662065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000127412</th>\n",
       "      <td>ENSG00000127412.6</td>\n",
       "      <td>7</td>\n",
       "      <td>142908101</td>\n",
       "      <td>142933808</td>\n",
       "      <td>142908101</td>\n",
       "      <td>142933808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000248569</th>\n",
       "      <td>ENSG00000248569.1</td>\n",
       "      <td>5</td>\n",
       "      <td>80351021</td>\n",
       "      <td>80351956</td>\n",
       "      <td>80351021</td>\n",
       "      <td>80351956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273797</th>\n",
       "      <td>ENSG00000273797.1</td>\n",
       "      <td>14</td>\n",
       "      <td>69617122</td>\n",
       "      <td>69617648</td>\n",
       "      <td>69617122</td>\n",
       "      <td>69617648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278771</th>\n",
       "      <td>ENSG00000278771.1</td>\n",
       "      <td>14</td>\n",
       "      <td>49853616</td>\n",
       "      <td>49853914</td>\n",
       "      <td>49853616</td>\n",
       "      <td>49853914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000252481</th>\n",
       "      <td>ENSG00000252481.1</td>\n",
       "      <td>14</td>\n",
       "      <td>95533355</td>\n",
       "      <td>95533629</td>\n",
       "      <td>95533355</td>\n",
       "      <td>95533629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000265315</th>\n",
       "      <td>ENSG00000265315.1</td>\n",
       "      <td>17</td>\n",
       "      <td>46537534</td>\n",
       "      <td>46537814</td>\n",
       "      <td>46537534</td>\n",
       "      <td>46537814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000206652</th>\n",
       "      <td>ENSG00000206652.1</td>\n",
       "      <td>1</td>\n",
       "      <td>16514122</td>\n",
       "      <td>16514285</td>\n",
       "      <td>16514122</td>\n",
       "      <td>16514285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277578</th>\n",
       "      <td>ENSG00000277578.1</td>\n",
       "      <td>15</td>\n",
       "      <td>84642487</td>\n",
       "      <td>84642763</td>\n",
       "      <td>84642487</td>\n",
       "      <td>84642763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000206737</th>\n",
       "      <td>ENSG00000206737.1</td>\n",
       "      <td>1</td>\n",
       "      <td>143729407</td>\n",
       "      <td>143729570</td>\n",
       "      <td>143729407</td>\n",
       "      <td>143729570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277925</th>\n",
       "      <td>ENSG00000277925.1</td>\n",
       "      <td>3</td>\n",
       "      <td>169764610</td>\n",
       "      <td>169765047</td>\n",
       "      <td>169764610</td>\n",
       "      <td>169765047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274266</th>\n",
       "      <td>ENSG00000274266.1</td>\n",
       "      <td>1</td>\n",
       "      <td>28507366</td>\n",
       "      <td>28507571</td>\n",
       "      <td>28507366</td>\n",
       "      <td>28507571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243650</th>\n",
       "      <td>ENSG00000243650.3</td>\n",
       "      <td>2</td>\n",
       "      <td>230507101</td>\n",
       "      <td>230507386</td>\n",
       "      <td>230507101</td>\n",
       "      <td>230507386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239002</th>\n",
       "      <td>ENSG00000239002.3</td>\n",
       "      <td>12</td>\n",
       "      <td>6510222</td>\n",
       "      <td>6510551</td>\n",
       "      <td>6510222</td>\n",
       "      <td>6510551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000266075</th>\n",
       "      <td>ENSG00000266075.2</td>\n",
       "      <td>1</td>\n",
       "      <td>3782815</td>\n",
       "      <td>3783111</td>\n",
       "      <td>3782815</td>\n",
       "      <td>3783111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000264501</th>\n",
       "      <td>ENSG00000264501.2</td>\n",
       "      <td>1</td>\n",
       "      <td>10306465</td>\n",
       "      <td>10306757</td>\n",
       "      <td>10306465</td>\n",
       "      <td>10306757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000241983</th>\n",
       "      <td>ENSG00000241983.3</td>\n",
       "      <td>19</td>\n",
       "      <td>39369153</td>\n",
       "      <td>39369448</td>\n",
       "      <td>39369153</td>\n",
       "      <td>39369448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000240877</th>\n",
       "      <td>ENSG00000240877.3</td>\n",
       "      <td>7</td>\n",
       "      <td>149125690</td>\n",
       "      <td>149125986</td>\n",
       "      <td>149125690</td>\n",
       "      <td>149125986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278700</th>\n",
       "      <td>ENSG00000278700.1</td>\n",
       "      <td>16</td>\n",
       "      <td>4658851</td>\n",
       "      <td>4659151</td>\n",
       "      <td>4658851</td>\n",
       "      <td>4659151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274012</th>\n",
       "      <td>ENSG00000274012.1</td>\n",
       "      <td>14</td>\n",
       "      <td>49862550</td>\n",
       "      <td>49862849</td>\n",
       "      <td>49862550</td>\n",
       "      <td>49862849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000252010</th>\n",
       "      <td>ENSG00000252010.1</td>\n",
       "      <td>2</td>\n",
       "      <td>233275727</td>\n",
       "      <td>233276002</td>\n",
       "      <td>233275727</td>\n",
       "      <td>233276002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000263740</th>\n",
       "      <td>ENSG00000263740.2</td>\n",
       "      <td>3</td>\n",
       "      <td>15738515</td>\n",
       "      <td>15738809</td>\n",
       "      <td>15738515</td>\n",
       "      <td>15738809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000202198</th>\n",
       "      <td>ENSG00000202198.1</td>\n",
       "      <td>6</td>\n",
       "      <td>52995620</td>\n",
       "      <td>52995950</td>\n",
       "      <td>52995620</td>\n",
       "      <td>52995950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000207110</th>\n",
       "      <td>ENSG00000207110.1</td>\n",
       "      <td>8</td>\n",
       "      <td>127999131</td>\n",
       "      <td>127999294</td>\n",
       "      <td>127999131</td>\n",
       "      <td>127999294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278233</th>\n",
       "      <td>ENSG00000278233.1</td>\n",
       "      <td>21</td>\n",
       "      <td>8212572</td>\n",
       "      <td>8212724</td>\n",
       "      <td>8212572</td>\n",
       "      <td>8212724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000222112</th>\n",
       "      <td>ENSG00000222112.1</td>\n",
       "      <td>1</td>\n",
       "      <td>33336566</td>\n",
       "      <td>33336864</td>\n",
       "      <td>33336566</td>\n",
       "      <td>33336864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000251791</th>\n",
       "      <td>ENSG00000251791.1</td>\n",
       "      <td>2</td>\n",
       "      <td>233288676</td>\n",
       "      <td>233288940</td>\n",
       "      <td>233288676</td>\n",
       "      <td>233288940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275215</th>\n",
       "      <td>ENSG00000275215.1</td>\n",
       "      <td>21</td>\n",
       "      <td>8395607</td>\n",
       "      <td>8395759</td>\n",
       "      <td>8395607</td>\n",
       "      <td>8395759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212443</th>\n",
       "      <td>ENSG00000212443.1</td>\n",
       "      <td>12</td>\n",
       "      <td>98599635</td>\n",
       "      <td>98599883</td>\n",
       "      <td>98599635</td>\n",
       "      <td>98599883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280502</th>\n",
       "      <td>ENSG00000280502.1</td>\n",
       "      <td>11</td>\n",
       "      <td>118994334</td>\n",
       "      <td>118994630</td>\n",
       "      <td>118994334</td>\n",
       "      <td>118994630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278189</th>\n",
       "      <td>ENSG00000278189.1</td>\n",
       "      <td>21</td>\n",
       "      <td>8439823</td>\n",
       "      <td>8439975</td>\n",
       "      <td>8439823</td>\n",
       "      <td>8439975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000206596</th>\n",
       "      <td>ENSG00000206596.1</td>\n",
       "      <td>14</td>\n",
       "      <td>34546714</td>\n",
       "      <td>34546877</td>\n",
       "      <td>34546714</td>\n",
       "      <td>34546877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000273541</th>\n",
       "      <td>ENSG00000273541.1</td>\n",
       "      <td>13</td>\n",
       "      <td>50104833</td>\n",
       "      <td>50105174</td>\n",
       "      <td>50104833</td>\n",
       "      <td>50105174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000262477</th>\n",
       "      <td>ENSG00000262477.1</td>\n",
       "      <td>18</td>\n",
       "      <td>32412182</td>\n",
       "      <td>32413236</td>\n",
       "      <td>32412182</td>\n",
       "      <td>32413236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17111 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_id_version chromosome  gene_start   gene_end  \\\n",
       "gene_id                                                                 \n",
       "ENSG00000277209   ENSG00000277209.1         14    20343075   20343407   \n",
       "ENSG00000266274   ENSG00000266274.2         17    30959564   30959850   \n",
       "ENSG00000263432   ENSG00000263432.2          5   123022487  123022783   \n",
       "ENSG00000239494   ENSG00000239494.2          1   169859756  169860052   \n",
       "ENSG00000212232   ENSG00000212232.1         20    17962710   17962946   \n",
       "ENSG00000211459   ENSG00000211459.2         MT         648       1601   \n",
       "ENSG00000210082   ENSG00000210082.2         MT        1671       3229   \n",
       "ENSG00000198888   ENSG00000198888.2         MT        3307       4262   \n",
       "ENSG00000198763   ENSG00000198763.3         MT        4470       5511   \n",
       "ENSG00000198804   ENSG00000198804.2         MT        5904       7445   \n",
       "ENSG00000228253   ENSG00000228253.1         MT        8366       8572   \n",
       "ENSG00000198899   ENSG00000198899.2         MT        8527       9207   \n",
       "ENSG00000198938   ENSG00000198938.2         MT        9207       9990   \n",
       "ENSG00000198840   ENSG00000198840.2         MT       10059      10404   \n",
       "ENSG00000212907   ENSG00000212907.2         MT       10470      10766   \n",
       "ENSG00000198886   ENSG00000198886.2         MT       10760      12137   \n",
       "ENSG00000198786   ENSG00000198786.2         MT       12337      14148   \n",
       "ENSG00000198695   ENSG00000198695.2         MT       14149      14673   \n",
       "ENSG00000198727   ENSG00000198727.2         MT       14747      15887   \n",
       "ENSG00000243339   ENSG00000243339.3          3   196399911  196400207   \n",
       "ENSG00000142530  ENSG00000142530.10         19    50466785   50476753   \n",
       "ENSG00000086967   ENSG00000086967.9         19    50432903   50466321   \n",
       "ENSG00000142552   ENSG00000142552.7         19    49527618   49546962   \n",
       "ENSG00000274029   ENSG00000274029.1         12   121190868  121191518   \n",
       "ENSG00000112379   ENSG00000112379.8          6   138161921  138344663   \n",
       "ENSG00000228150   ENSG00000228150.1          1     9942923    9949974   \n",
       "ENSG00000233705   ENSG00000233705.6          7   107653968  107662151   \n",
       "ENSG00000127412   ENSG00000127412.6          7   142908101  142933808   \n",
       "ENSG00000248569   ENSG00000248569.1          5    80351021   80351956   \n",
       "ENSG00000273797   ENSG00000273797.1         14    69617122   69617648   \n",
       "...                             ...        ...         ...        ...   \n",
       "ENSG00000278771   ENSG00000278771.1         14    49853616   49853914   \n",
       "ENSG00000252481   ENSG00000252481.1         14    95533355   95533629   \n",
       "ENSG00000265315   ENSG00000265315.1         17    46537534   46537814   \n",
       "ENSG00000206652   ENSG00000206652.1          1    16514122   16514285   \n",
       "ENSG00000277578   ENSG00000277578.1         15    84642487   84642763   \n",
       "ENSG00000206737   ENSG00000206737.1          1   143729407  143729570   \n",
       "ENSG00000277925   ENSG00000277925.1          3   169764610  169765047   \n",
       "ENSG00000274266   ENSG00000274266.1          1    28507366   28507571   \n",
       "ENSG00000243650   ENSG00000243650.3          2   230507101  230507386   \n",
       "ENSG00000239002   ENSG00000239002.3         12     6510222    6510551   \n",
       "ENSG00000266075   ENSG00000266075.2          1     3782815    3783111   \n",
       "ENSG00000264501   ENSG00000264501.2          1    10306465   10306757   \n",
       "ENSG00000241983   ENSG00000241983.3         19    39369153   39369448   \n",
       "ENSG00000240877   ENSG00000240877.3          7   149125690  149125986   \n",
       "ENSG00000278700   ENSG00000278700.1         16     4658851    4659151   \n",
       "ENSG00000274012   ENSG00000274012.1         14    49862550   49862849   \n",
       "ENSG00000252010   ENSG00000252010.1          2   233275727  233276002   \n",
       "ENSG00000263740   ENSG00000263740.2          3    15738515   15738809   \n",
       "ENSG00000202198   ENSG00000202198.1          6    52995620   52995950   \n",
       "ENSG00000207110   ENSG00000207110.1          8   127999131  127999294   \n",
       "ENSG00000278233   ENSG00000278233.1         21     8212572    8212724   \n",
       "ENSG00000222112   ENSG00000222112.1          1    33336566   33336864   \n",
       "ENSG00000251791   ENSG00000251791.1          2   233288676  233288940   \n",
       "ENSG00000275215   ENSG00000275215.1         21     8395607    8395759   \n",
       "ENSG00000212443   ENSG00000212443.1         12    98599635   98599883   \n",
       "ENSG00000280502   ENSG00000280502.1         11   118994334  118994630   \n",
       "ENSG00000278189   ENSG00000278189.1         21     8439823    8439975   \n",
       "ENSG00000206596   ENSG00000206596.1         14    34546714   34546877   \n",
       "ENSG00000273541   ENSG00000273541.1         13    50104833   50105174   \n",
       "ENSG00000262477   ENSG00000262477.1         18    32412182   32413236   \n",
       "\n",
       "                 transcript_start  transcript_end  \n",
       "gene_id                                            \n",
       "ENSG00000277209          20343075        20343407  \n",
       "ENSG00000266274          30959564        30959850  \n",
       "ENSG00000263432         123022487       123022783  \n",
       "ENSG00000239494         169859756       169860052  \n",
       "ENSG00000212232          17962710        17962946  \n",
       "ENSG00000211459               648            1601  \n",
       "ENSG00000210082              1671            3229  \n",
       "ENSG00000198888              3307            4262  \n",
       "ENSG00000198763              4470            5511  \n",
       "ENSG00000198804              5904            7445  \n",
       "ENSG00000228253              8366            8572  \n",
       "ENSG00000198899              8527            9207  \n",
       "ENSG00000198938              9207            9990  \n",
       "ENSG00000198840             10059           10404  \n",
       "ENSG00000212907             10470           10766  \n",
       "ENSG00000198886             10760           12137  \n",
       "ENSG00000198786             12337           14148  \n",
       "ENSG00000198695             14149           14673  \n",
       "ENSG00000198727             14747           15887  \n",
       "ENSG00000243339         196399911       196400207  \n",
       "ENSG00000142530          50466785        50475485  \n",
       "ENSG00000086967          50432903        50466321  \n",
       "ENSG00000142552          49527618        49543638  \n",
       "ENSG00000274029         121190868       121191518  \n",
       "ENSG00000112379         138161921       138344663  \n",
       "ENSG00000228150           9942923         9949974  \n",
       "ENSG00000233705         107653968       107662065  \n",
       "ENSG00000127412         142908101       142933808  \n",
       "ENSG00000248569          80351021        80351956  \n",
       "ENSG00000273797          69617122        69617648  \n",
       "...                           ...             ...  \n",
       "ENSG00000278771          49853616        49853914  \n",
       "ENSG00000252481          95533355        95533629  \n",
       "ENSG00000265315          46537534        46537814  \n",
       "ENSG00000206652          16514122        16514285  \n",
       "ENSG00000277578          84642487        84642763  \n",
       "ENSG00000206737         143729407       143729570  \n",
       "ENSG00000277925         169764610       169765047  \n",
       "ENSG00000274266          28507366        28507571  \n",
       "ENSG00000243650         230507101       230507386  \n",
       "ENSG00000239002           6510222         6510551  \n",
       "ENSG00000266075           3782815         3783111  \n",
       "ENSG00000264501          10306465        10306757  \n",
       "ENSG00000241983          39369153        39369448  \n",
       "ENSG00000240877         149125690       149125986  \n",
       "ENSG00000278700           4658851         4659151  \n",
       "ENSG00000274012          49862550        49862849  \n",
       "ENSG00000252010         233275727       233276002  \n",
       "ENSG00000263740          15738515        15738809  \n",
       "ENSG00000202198          52995620        52995950  \n",
       "ENSG00000207110         127999131       127999294  \n",
       "ENSG00000278233           8212572         8212724  \n",
       "ENSG00000222112          33336566        33336864  \n",
       "ENSG00000251791         233288676       233288940  \n",
       "ENSG00000275215           8395607         8395759  \n",
       "ENSG00000212443          98599635        98599883  \n",
       "ENSG00000280502         118994334       118994630  \n",
       "ENSG00000278189           8439823         8439975  \n",
       "ENSG00000206596          34546714        34546877  \n",
       "ENSG00000273541          50104833        50105174  \n",
       "ENSG00000262477          32412182        32413236  \n",
       "\n",
       "[17111 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_chrom.chromosome.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(gene_chrom['chromosome'].value_counts().index))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tree-map images generation\n",
    "\n",
    "The generation of these images is going to be done by the execution of a RScript: 'Gene_chromosome_treemap_generator.R'. The function 'treemap' from 'treemap' package is used in it. To execute this RScript, it is required a csv file (exp_to_chromosome_tree_map.csv) being in the folder containing the script. This csv file contains the gene expression profiles and the chromosomic information for generating the images.\n",
    "\n",
    "One image corresponds to one patient, and each gene is represented by a square colored depending on its expression level. Genes belonging to the same chromosome are placed in the same area of the tree-map image. Inside these areas, in which all genes belong to the same chromosome, genes are sorted by the mean expression it has through all patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_chrom.rename(columns={'gene_id_version': 'ensId'}, inplace=True)\n",
    "exp_to_tree_map = pd.concat([non_brca_ex, brca_ex]).T\n",
    "exp_to_tree_map['ensId'] = exp_to_tree_map.index\n",
    "exp_to_tree_map['tamPixel'] = np.ones(exp_to_tree_map.shape[0])\n",
    "exp_to_tree_map['order'] = pd.concat([non_brca_ex, brca_ex]).mean() # order genes in the map by mean expression level\n",
    "exp_to_tree_map = pd.merge(gene_chrom, exp_to_tree_map, on='ensId')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17111, 10543)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_to_tree_map.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('genes_remaining.txt', 'w') as f:\n",
    "    f.write(str(set(exp_to_tree_map.ensId)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.3 s, sys: 384 ms, total: 11.6 s\n",
      "Wall time: 11.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "exp_to_tree_map.to_csv('exp_to_chromosome_tree_map.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "source('Gene_chromosome_treemap_generator.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "# DL with Python - François Chollet (pp. 162-163)\n",
    "from PIL import Image\n",
    "from keras.preprocessing import image\n",
    "import glob\n",
    "\n",
    "path = '../data/CHR_TREEMAPS/'\n",
    "filelist = glob.glob(path + '*.png')\n",
    "\n",
    "brca_images = dict()\n",
    "non_brca_images = dict()\n",
    "w, h = (480, 480)\n",
    "basewidth = 100\n",
    "wpercent = (basewidth/float(w))\n",
    "hsize = int((float(h)*float(wpercent)))\n",
    "i = 0 \n",
    "for filename in filelist:\n",
    "    sample = filename.split(\" \")[1]\n",
    "    #print(filename)\n",
    "    if(True):\n",
    "        if('NON' in sample):\n",
    "            #print(filename)\n",
    "            #i+=1\n",
    "            #print(i)\n",
    "            non_brca_images[sample.split(\"_\")[0].replace('.','-')] = image.load_img(filename).crop((11, 11, w-13, h-13)).resize((basewidth,hsize), Image.ANTIALIAS)\n",
    "        else:\n",
    "            #print(filename)\n",
    "            #i+=1\n",
    "            #print(i)\n",
    "            brca_images[sample.split(\"_\")[0].replace('.','-')] = image.load_img(filename).crop((11, 11, w-13, h-13)).resize((basewidth,hsize), Image.ANTIALIAS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_images[list(non_brca_images.keys())[11]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "brca_clinical = pd.read_hdf('data/TCGA_data.h5', key='brca_clinical')\n",
    "non_brca_clinical = pd.read_hdf('data/TCGA_data.h5', key='non_brca_clinical')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deleting samples with 'NaN's in OS-columns \n",
    "\n",
    "## non_brca\n",
    "non_brca_clinical['OS.time'] = non_brca_clinical['OS.time'].map(lambda x: np.nan if x == 'NaN' else x)\n",
    "non_brca_clinical['OS'] = non_brca_clinical['OS'].map(lambda x: np.nan if x == 'NaN' else x)\n",
    "non_brca_clinical.dropna(subset=['OS.time', 'OS'], inplace=True)\n",
    "non_brca_clinical['OS.time'] = non_brca_clinical['OS.time'].astype(float)\n",
    "non_brca_clinical['OS'] = non_brca_clinical['OS'].astype(float)\n",
    "\n",
    "## brca\n",
    "brca_clinical['OS.time'] = brca_clinical['OS.time'].map(lambda x: np.nan if x == 'NaN' else x)\n",
    "brca_clinical['OS'] = brca_clinical['OS'].map(lambda x: np.nan if x == 'NaN' else x)\n",
    "brca_clinical.dropna(subset=['OS.time', 'OS'], inplace=True)\n",
    "brca_clinical['OS.time'] = brca_clinical['OS.time'].astype(float)\n",
    "brca_clinical['OS'] = brca_clinical['OS'].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_clinical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_brca = set(list(brca_images.keys())).intersection(brca_clinical.index)\n",
    "for sample in list(brca_images.keys()):\n",
    "    if(sample not in common_brca): del brca_images[sample] \n",
    "brca_clinical = brca_clinical.loc[common_brca]\n",
    "\n",
    "common_non_brca = set(list(non_brca_images.keys())).intersection(non_brca_clinical.index)\n",
    "for sample in list(non_brca_images.keys()):\n",
    "    if(sample not in common_non_brca): del non_brca_images[sample] \n",
    "non_brca_clinical = non_brca_clinical.loc[common_non_brca]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create 39 equally spaced intervals for follow-up time, from 0 to 5 years:\n",
    "breaks = np.arange(0.,365.*5,365./8)\n",
    "n_intervals = len(breaks)-1\n",
    "timegap = breaks[1:] - breaks[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import nnet_survival\n",
    "non_brca_y = nnet_survival.make_surv_array(non_brca_clinical['OS.time'],\n",
    "                             non_brca_clinical['OS'],\n",
    "                             breaks)\n",
    "brca_y = nnet_survival.make_surv_array(brca_clinical['OS.time'],\n",
    "                             brca_clinical['OS'],\n",
    "                             breaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_clinical.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_images.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_images[non_brca_clinical.index[100]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_images[list(non_brca_images.keys())[100]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "breaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_brca_y[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "import numpy as np\n",
    "samples = list(non_brca_clinical.index)\n",
    "\n",
    "X_non_brca = np.empty((len(samples), 100, 100, 3))\n",
    "\n",
    "for i in range(0, len(samples)):\n",
    "    X_non_brca[i] = image.img_to_array(non_brca_images[samples[i]])\n",
    "#img = non_brca_images[list(non_brca_images.keys())[9]]\n",
    "#x = image.img_to_array(img)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_non_brca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del non_brca_images\n",
    "del brca_images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# Split dataset into train and test\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "X_non_brca_train, X_non_brca_test, y_non_brca_train, y_non_brca_test, train_index, test_index = train_test_split(X_non_brca, non_brca_y, np.arange(len(X_non_brca)), test_size=0.2, random_state=42)\n",
    "X_non_brca_train, X_non_brca_val, y_non_brca_train, y_non_brca_val, train_index, val_index = train_test_split(X_non_brca_train, y_non_brca_train, train_index, test_size=0.2, random_state=44)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_non_brca_train = np.array(y_non_brca_train, dtype=int)\n",
    "y_non_brca_test = np.array(y_non_brca_test, dtype=int)\n",
    "y_non_brca_val = np.array(y_non_brca_val, dtype=int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "del X_non_brca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Dataset size:\", X_non_brca_train.shape, X_non_brca_test.shape, X_non_brca_val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import keras.backend as K\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Input, Dense, Conv2D, MaxPool2D, Activation, Dropout, Flatten\n",
    "from keras import regularizers, optimizers\n",
    "from keras.layers.normalization import BatchNormalization\n",
    "from keras.callbacks import EarlyStopping\n",
    "import warnings\n",
    "from keras.regularizers import l1\n",
    "from keras.optimizers import Adam\n",
    "import tensorflow as tf"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "K.clear_session()\n",
    "\n",
    "config = tf.ConfigProto()\n",
    "config.gpu_options.allow_growth = True\n",
    "config.gpu_options.per_process_gpu_memory_fraction = 0.4\n",
    "\n",
    "K.set_session(tf.Session(config=config))\n",
    "\n",
    "model = Sequential()\n",
    "\n",
    "# Conv layer 1 ---\n",
    "model.add(Conv2D(32, kernel_size=(3,3),\n",
    "                input_shape = (X_non_brca_train.shape[1], X_non_brca_train.shape[2], X_non_brca_train.shape[3])))\n",
    "model.add(BatchNormalization())\n",
    "model.add(Activation('relu'))\n",
    "model.add(MaxPool2D(pool_size=(2,2)))\n",
    "model.add(Dropout(0.25))\n",
    "# Conv layer 2 ---\n",
    "model.add(Conv2D(64, kernel_size=(3,3)))\n",
    "model.add(BatchNormalization())\n",
    "model.add(Activation('relu'))\n",
    "model.add(MaxPool2D(pool_size=(2,2)))\n",
    "model.add(Dropout(0.25))\n",
    "# # Conv layer 3 ---\n",
    "model.add(Conv2D(128, kernel_size=(3,3)))\n",
    "model.add(BatchNormalization())\n",
    "model.add(Activation('relu'))\n",
    "model.add(MaxPool2D(pool_size=(2,2)))\n",
    "model.add(Dropout(0.25))\n",
    "# Flatten layer ---\n",
    "model.add(Flatten())\n",
    "#model.add(Dense(36864, activation='relu'))\n",
    "# Dense layer 1 ---\n",
    "model.add(Dense(128))\n",
    "model.add(BatchNormalization())\n",
    "model.add(Activation('relu'))\n",
    "model.add(Dropout(0.5))\n",
    "# Dense layer 2 ---\n",
    "model.add(Dense(64))\n",
    "model.add(BatchNormalization())\n",
    "model.add(Activation('relu'))\n",
    "# Dense layer 3 ---\n",
    "# model.add(Dense(256))\n",
    "# model.add(BatchNormalization())\n",
    "# model.add(Activation('relu'))\n",
    "# Output layer ---\n",
    "model.add(Dense(n_intervals, activation='sigmoid'))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "optimizer = Adam(lr=0.001)\n",
    "\n",
    "model.compile(loss=nnet_survival.surv_likelihood(n_intervals), optimizer=optimizer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using C-index as evaluation metric (using a custom callback)\n",
    "\n",
    "from lifelines.utils import concordance_index\n",
    "from keras.callbacks import Callback\n",
    "\n",
    "class CIndex(Callback):\n",
    "    \"\"\"\n",
    "    Callback that computes the C-index metric both on training and test data after each epoch.\n",
    "    \n",
    "    Arguments:\n",
    "        train_x: Numpy array containing the training dataset.\n",
    "        train_time: Numpy array with survival time for training dataset.\n",
    "        train_event: Numpy array with survival event for training dataset.\n",
    "        val_x: Numpy array containing the test dataset.\n",
    "        val_time: Numpy array with survival time for test dataset.\n",
    "        val_event: Numpy array with survival event for test dataset.\n",
    "        filepath: path to save the model.\n",
    "        years: number of years to compute c-index\n",
    "        patience: patience for early-stop training\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, train_x, train_time, train_event, val_x, val_time, val_event, filepath, years=2, patience=5):\n",
    "        super(Callback, self).__init__()\n",
    "        self.X_tr = train_x\n",
    "        self.time_tr = train_time\n",
    "        self.event_tr = train_event\n",
    "        self.X_val = val_x\n",
    "        self.time_val = val_time\n",
    "        self.event_val = val_event\n",
    "        self.years = years\n",
    "        self.patience = patience\n",
    "        self.best = 0.0\n",
    "        self.wait = 0  #counter for patience\n",
    "        self.best_rounds = 1\n",
    "        self.counter = 0\n",
    "        self.filepath = filepath\n",
    "        \n",
    "    def on_train_begin(self, logs={}):\n",
    "        self.c_index_dict = {'c-index': [], 'val_c-index': []}\n",
    "\n",
    "    def on_epoch_end(self, epoch, logs={}):\n",
    "        self.counter +=1\n",
    "\n",
    "        y_tr=self.model.predict_proba(self.X_tr,verbose=0)\n",
    "        yr_surv=np.cumprod(y_tr[:,0:np.nonzero(breaks>365*self.years)[0][0]], axis=1)[:,-1]\n",
    "        c_index_tr = concordance_index(self.time_tr, yr_surv, self.event_tr)\n",
    "        y_val=self.model.predict_proba(self.X_val,verbose=0)\n",
    "        yr_surv_val=np.cumprod(y_val[:,0:np.nonzero(breaks>365*self.years)[0][0]], axis=1)[:,-1]\n",
    "        c_index_val = concordance_index(self.time_val, yr_surv_val, self.event_val)\n",
    "\n",
    "        print('\\tc-index: %s - val_c-index: %s' % (str(round(c_index_tr, 4)), str(round(c_index_val, 4))))\n",
    "        \n",
    "        if (c_index_val > self.best):\n",
    "            self.c_index_dict['c-index'].append(c_index_tr)\n",
    "            self.c_index_dict['val_c-index'].append(c_index_val)\n",
    "\n",
    "            self.best = c_index_val\n",
    "            self.best_rounds = self.counter\n",
    "            self.wait = 0\n",
    "            self.model.save_weights(self.filepath, overwrite=True)\n",
    "        else:\n",
    "            if self.wait >= self.patience: #no more patience, retrieve best model\n",
    "                self.model.stop_training = True\n",
    "                print('Best number of rounds: %d \\nValidation C-Index: %f \\n' % (self.best_rounds, self.best))\n",
    "                self.model.load_weights(self.filepath)\n",
    "            self.wait += 1"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "%time\n",
    "\n",
    "hist_c_index = CIndex(X_non_brca_train,\n",
    "                    non_brca_clinical.iloc[train_index]['OS.time'],\n",
    "                    non_brca_clinical.iloc[train_index]['OS'],\n",
    "                    X_non_brca_val,\n",
    "                    non_brca_clinical.iloc[val_index]['OS.time'],\n",
    "                    non_brca_clinical.iloc[val_index]['OS'],\n",
    "                    filepath = 'keras-models/best.h5')\n",
    "\n",
    "#callbacks = [EarlyStopping(monitor='loss', min_delta=0.01, patience=2, verbose=2, mode=\"min\")]\n",
    "callbacks = [hist_c_index]\n",
    "\n",
    "\n",
    "batch = 128\n",
    "n_epoch = 50\n",
    "\n",
    "history = model.fit(X_non_brca_train, \n",
    "                    y_non_brca_train, \n",
    "                    batch_size=batch, \n",
    "                    epochs=n_epoch, \n",
    "                    callbacks=callbacks, \n",
    "                    validation_data = [X_non_brca_val, y_non_brca_val])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from hyperopt import hp, fmin, tpe, hp, STATUS_OK, Trials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "space = {\n",
    "    'conv_choice': hp.choice('conv_num_layers', \n",
    "                            [{'layers': 'two',\n",
    "                             'units_2conv2': hp.choice('units_2conv2', [16, 32, 64, 128]),\n",
    "                              '2kernel2': hp.choice('2kernel2', [3,5]),\n",
    "                              '2pool2': 2,\n",
    "                             'dropout_2conv2': hp.choice('dropout_2conv2', [0.25, 0.5, 0.75])\n",
    "                            },\n",
    "                            {'layers': 'three',\n",
    "                             'units_3conv2': hp.choice('units_3conv2', [16, 32, 64, 128]),\n",
    "                             '3kernel2': hp.choice('3kernel2', [3,5]),\n",
    "                             '3pool2': 2,\n",
    "                             'dropout_3conv2': hp.choice('dropout_3conv2', [0.25, 0.5, 0.75]),\n",
    "                             'units_3conv3': hp.choice('units_3conv3', [16, 32, 64, 128]),\n",
    "                             '3kernel3': hp.choice('kernel3', [3,5]),\n",
    "                             '3pool3': 2,\n",
    "                             'dropout_3conv3': hp.choice('dropout_3conv3', [0.25, 0.5, 0.75])\n",
    "                            }]\n",
    "                            ),\n",
    "    'dense_choice': hp.choice('dense_num_layers', \n",
    "                            [{'layers': 'two',\n",
    "                             'units_2dense2': hp.choice('units_2dense2', [16, 32, 64, 128]),\n",
    "                             'dropout_2dense2': hp.choice('dropout_2dense2', [0.25, 0.5, 0.75])\n",
    "                            },\n",
    "                            {'layers': 'three',\n",
    "                             'units_3dense2': hp.choice('units_3dense2', [16, 32, 64, 128]),\n",
    "                             'dropout_3dense2': hp.choice('dropout_3dense2', [0.25, 0.5, 0.75]),\n",
    "                             'units_3dense3': hp.choice('units_3dense3', [16, 32, 64, 128]),\n",
    "                             'dropout_3dense3': hp.choice('dropout_3dense3', [0.25, 0.5, 0.75])\n",
    "                            }]\n",
    "                            ),\n",
    "    'units_conv1': hp.choice('units_conv1', [16, 32, 64, 128]),\n",
    "    'kernel1': hp.choice('kernel1', [3,5]),\n",
    "    'pool1': 2,\n",
    "    'dropout_conv1': hp.choice('dropout_conv1', [0.25, 0.5, 0.75]),\n",
    "    'units_dense1': hp.choice('units_dense1', [16, 32, 64, 128]),\n",
    "    'dropout_dense1': hp.choice('dropout_dense1', [0.25, 0.5, 0.75]),\n",
    "    'batch_size' : hp.choice('batch_size', [16, 32, 64, 128]),\n",
    "    'nb_epochs' : 100,\n",
    "    'lr': hp.loguniform('lr', np.log(0.0001), np.log(0.01)),\n",
    "    'activation': 'relu'                                            \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train(params):\n",
    "    print('Params testing: ', params)\n",
    "    print('\\n')\n",
    "\n",
    "    # Prevent GPU memory allocation problems\n",
    "    config = tf.ConfigProto()\n",
    "    config.gpu_options.allow_growth = True\n",
    "    config.gpu_options.per_process_gpu_memory_fraction = 0.4\n",
    "    K.set_session(tf.Session(config=config))\n",
    "\n",
    "    ## CREATE MODEL\n",
    "    model = Sequential()\n",
    "\n",
    "    # Conv layer 1 ---\n",
    "    model.add(Conv2D(filters=params['units_conv1'], \n",
    "                     kernel_size=(params['kernel1'], params['kernel1']),\n",
    "                     input_shape = (X_non_brca_train.shape[1],\n",
    "                                    X_non_brca_train.shape[2],\n",
    "                                    X_non_brca_train.shape[3])))\n",
    "    model.add(BatchNormalization())\n",
    "    model.add(Activation(params['activation']))\n",
    "    model.add(MaxPool2D(pool_size=(params['pool1'],params['pool1'])))\n",
    "    model.add(Dropout(params['dropout_conv1']))\n",
    "\n",
    "    if params['conv_choice']['layers'] == 'two':\n",
    "        # Conv layer 2 ---\n",
    "        model.add(Conv2D(filters=params['conv_choice']['units_2conv2'], \n",
    "                     kernel_size=(params['conv_choice']['2kernel2'], params['conv_choice']['2kernel2'])))\n",
    "        model.add(BatchNormalization())\n",
    "        model.add(Activation(params['activation']))\n",
    "        model.add(MaxPool2D(pool_size=(params['conv_choice']['2pool2'], params['conv_choice']['2pool2'])))\n",
    "        model.add(Dropout(params['conv_choice']['dropout_2conv2']))\n",
    "    elif params['conv_choice']['layers'] == 'three':\n",
    "        # Conv layer 2 ---\n",
    "        model.add(Conv2D(filters=params['conv_choice']['units_3conv2'], \n",
    "                     kernel_size=(params['conv_choice']['3kernel2'], params['conv_choice']['3kernel2'])))\n",
    "        model.add(BatchNormalization())\n",
    "        model.add(Activation(params['activation']))\n",
    "        model.add(MaxPool2D(pool_size=(params['conv_choice']['3pool2'], params['conv_choice']['3pool2'])))\n",
    "        model.add(Dropout(params['conv_choice']['dropout_3conv2']))\n",
    "        # Conv layer 3 ---\n",
    "        model.add(Conv2D(filters=params['conv_choice']['units_3conv3'], \n",
    "                     kernel_size=(params['conv_choice']['3kernel3'], params['conv_choice']['3kernel3'])))\n",
    "        model.add(BatchNormalization())\n",
    "        model.add(Activation(params['activation']))\n",
    "        model.add(MaxPool2D(pool_size=(params['conv_choice']['3pool3'], params['conv_choice']['3pool3'])))\n",
    "        model.add(Dropout(params['conv_choice']['dropout_3conv3']))\n",
    "\n",
    "    # Flatten layer ---\n",
    "    model.add(Flatten())\n",
    "    \n",
    "    # Dense layer 1 ---\n",
    "    model.add(Dense(params['units_dense1']))\n",
    "    model.add(BatchNormalization())\n",
    "    model.add(Activation(params['activation']))\n",
    "    model.add(Dropout(params['dropout_dense1']))\n",
    "\n",
    "    if params['dense_choice']['layers']== 'two':\n",
    "        # Dense layer 2 ---\n",
    "        model.add(Dense(params['dense_choice']['units_2dense2']))\n",
    "        model.add(BatchNormalization())\n",
    "        model.add(Activation(params['activation']))\n",
    "        model.add(Dropout(params['dense_choice']['dropout_2dense2']))\n",
    "    elif params['dense_choice']['layers']== 'three':\n",
    "        # Dense layer 2 ---\n",
    "        model.add(Dense(params['dense_choice']['units_3dense2']))\n",
    "        model.add(BatchNormalization())\n",
    "        model.add(Activation(params['activation']))\n",
    "        model.add(Dropout(params['dense_choice']['dropout_3dense2']))\n",
    "        # Dense layer 3 ---\n",
    "        model.add(Dense(params['dense_choice']['units_3dense3']))\n",
    "        model.add(BatchNormalization())\n",
    "        model.add(Activation(params['activation']))\n",
    "        model.add(Dropout(params['dense_choice']['dropout_3dense3']))\n",
    "        \n",
    "    # Output layer ---\n",
    "    model.add(Dense(n_intervals, activation='sigmoid'))\n",
    "\n",
    "    optimizer = Adam(lr=params['lr'])\n",
    "\n",
    "    model.compile(loss=nnet_survival.surv_likelihood(n_intervals), optimizer=optimizer)\n",
    "    \n",
    "    model.summary()\n",
    "    \n",
    "    batch = params['batch_size']\n",
    "    n_epoch = params['nb_epochs']\n",
    "\n",
    "    hist_c_index = CIndex(X_non_brca_train,\n",
    "                    non_brca_clinical.iloc[train_index]['OS.time'],\n",
    "                    non_brca_clinical.iloc[train_index]['OS'],\n",
    "                    X_non_brca_val,\n",
    "                    non_brca_clinical.iloc[val_index]['OS.time'],\n",
    "                    non_brca_clinical.iloc[val_index]['OS'],\n",
    "                    filepath = 'keras-models/best.h5')\n",
    "\n",
    "    callbacks = [hist_c_index]    \n",
    "    \n",
    "    ## Fit model using training data\n",
    "    start = timer()\n",
    "    model.fit(X_non_brca_train, \n",
    "              y_non_brca_train, \n",
    "              batch_size=batch, \n",
    "              epochs=n_epoch, \n",
    "              validation_data = [X_non_brca_val, y_non_brca_val],\n",
    "              callbacks=callbacks)\n",
    "    run_time = timer() - start\n",
    "    \n",
    "    predictions = model.predict_proba(X_non_brca_test,verbose=0)\n",
    "    twoyr_surv=np.cumprod(predictions[:,0:np.nonzero(breaks>365*2)[0][0]], axis=1)[:,-1]\n",
    "    score = concordance_index(non_brca_clinical.iloc[test_index]['OS.time'], twoyr_surv, non_brca_clinical.iloc[test_index]['OS'])\n",
    "\n",
    "    # Return loss\n",
    "    return {'loss': score*(-1),\n",
    "            'params': params,\n",
    "            'train_time': run_time,\n",
    "            'status': STATUS_OK, \n",
    "            'rounds':hist_c_index.best_rounds}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from timeit import default_timer as timer\n",
    "from multiprocessing import Pool\n",
    "from hyperopt import STATUS_OK\n",
    "\n",
    "def objective(params):\n",
    "    from sklearn.metrics import roc_auc_score\n",
    "    \n",
    "    from keras.models import Sequential\n",
    "    from keras.layers import Input, Dense, Conv1D, MaxPool1D, Activation, Dropout, Flatten\n",
    "    from keras import regularizers, optimizers\n",
    "    from keras.layers.normalization import BatchNormalization\n",
    "    from keras.callbacks import EarlyStopping\n",
    "\n",
    "    from keras.regularizers import l1\n",
    "    from keras.optimizers import Adam\n",
    "   \n",
    "    with Pool(1) as p:\n",
    "            trial = p.apply(train, args=(params, ))\n",
    "\n",
    "    return trial"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "%%time\n",
    "\n",
    "trials_non_brca = Trials()\n",
    "\n",
    "best_non_brca = fmin(objective, space, algo=tpe.suggest, trials=trials_non_brca, max_evals=50)\n",
    "\n",
    "print (best_non_brca)\n",
    "print (trials_non_brca.best_trial)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import pickle\n",
    "\n",
    "with open(\"hyper_opt_chromosome.pkl\", 'wb') as f:\n",
    "    pickle.dump([trials_non_brca, best_non_brca], f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open(\"hyper_opt_chromosome.pkl\", 'rb') as f:\n",
    "    [trials_non_brca, best_non_brca] = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_non_brca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trials_non_brca.best_trial['result']['params']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "space = {\n",
    "    'conv_choice': hp.choice('conv_num_layers', \n",
    "                            [{'layers': 'two',\n",
    "                             'units_2conv2': hp.choice('units_2conv2', [16, 32, 64, 128]),\n",
    "                              '2kernel2': hp.choice('2kernel2', [3,5]),\n",
    "                              '2pool2': 2,\n",
    "                             'dropout_2conv2': hp.choice('dropout_2conv2', [0.25, 0.5, 0.75])\n",
    "                            },\n",
    "                            {'layers': 'three',\n",
    "                             'units_3conv2': hp.choice('units_3conv2', [16, 32, 64, 128]),\n",
    "                             '3kernel2': hp.choice('3kernel2', [3,5]),\n",
    "                             '3pool2': 2,\n",
    "                             'dropout_3conv2': hp.choice('dropout_3conv2', [0.25, 0.5, 0.75]),\n",
    "                             'units_3conv3': hp.choice('units_3conv3', [16, 32, 64, 128]),\n",
    "                             '3kernel3': hp.choice('kernel3', [3,5]),\n",
    "                             '3pool3': 2,\n",
    "                             'dropout_3conv3': hp.choice('dropout_3conv3', [0.25, 0.5, 0.75])\n",
    "                            }]\n",
    "                            ),\n",
    "    'dense_choice': hp.choice('dense_num_layers', \n",
    "                            [{'layers': 'two',\n",
    "                             'units_2dense2': hp.choice('units_2dense2', [16, 32, 64, 128]),\n",
    "                             'dropout_2dense2': hp.choice('dropout_2dense2', [0.25, 0.5, 0.75])\n",
    "                            },\n",
    "                            {'layers': 'three',\n",
    "                             'units_3dense2': hp.choice('units_3dense2', [16, 32, 64, 128]),\n",
    "                             'dropout_3dense2': hp.choice('dropout_3dense2', [0.25, 0.5, 0.75]),\n",
    "                             'units_3dense3': hp.choice('units_3dense3', [16, 32, 64, 128]),\n",
    "                             'dropout_3dense3': hp.choice('dropout_3dense3', [0.25, 0.5, 0.75])\n",
    "                            }]\n",
    "                            ),\n",
    "    'units_conv1': hp.choice('units_conv1', [16, 32, 64, 128]),\n",
    "    'kernel1': hp.choice('kernel1', [3,5]),\n",
    "    'pool1': 2,\n",
    "    'dropout_conv1': hp.choice('dropout_conv1', [0.25, 0.5, 0.75]),\n",
    "    'units_dense1': hp.choice('units_dense1', [16, 32, 64, 128]),\n",
    "    'dropout_dense1': hp.choice('dropout_dense1', [0.25, 0.5, 0.75]),\n",
    "    'batch_size' : hp.choice('batch_size', [16, 32, 64, 128]),\n",
    "    'nb_epochs' : 100,\n",
    "    'lr': hp.loguniform('lr', np.log(0.0001), np.log(0.01)),\n",
    "    'activation': 'relu'                                            \n",
    "}\n",
    "best_non_brca['units_conv1'] = 32\n",
    "best_non_brca['kernel1'] = 3\n",
    "best_non_brca['activation'] = 'relu'\n",
    "\n",
    "best_non_brca['pool1'] = 2\n",
    "best_non_brca['dropout_conv1'] = 0.5\n",
    "best_non_brca['conv_choice'] = {'layers' : 'two', \n",
    "                                'units_2conv2' : 64,\n",
    "                                '2kernel2' : 5,\n",
    "                                '2pool2' : 2,\n",
    "                                'dropout_2conv2' : 0.5}\n",
    "\n",
    "best_non_brca['dropout_dense1'] = 0.5\n",
    "best_non_brca['units_dense1'] = 32\n",
    "best_non_brca['dense_choice'] = {'layers' : 'three',\n",
    "                                'units_3dense2' : 128,\n",
    "                                'units_3dense3' : 32,\n",
    "                                'dropout_3dense2' : 0.75,\n",
    "                                'dropout_3dense3' : 0.25}\n",
    "\n",
    "best_non_brca['batch_size'] = 64\n",
    "best_non_brca['nb_epochs'] = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "objective(trials_non_brca.best_trial['result']['params'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
